### ISSUE TITLE
Consider allowing a worker to consume from multiple queues

### ISSUE THREAD

**Issue by barbolo** [jondot/sneakers#298]
Is it possible to specify a worker that pulls jobs from multiple queues?

I'm looking for a way to load balance jobs from multiple queues, but to also be able to use all processing resources for a single queue  if all other queues are empty.

**Comment by michaelklishin** [comment]
A single worker currently consumes from a single queue. You can work around this by creating a few workers that share pretty much everything besides their source queue (`from_queue` in the DSL).

**Comment by barbolo** [comment]
Michael, the problem with the setting you've suggested is that you may end up with idle threads and limited capacity per worker (queue).

Can I pull request a change to support multiple queues? Is that coherent with the project goals?

**Comment by michaelklishin** [comment]
While having 1 worker instead of N can mean that you'd likely have more total threads used (although it's up to you how many you configure per worker), I don't see it as a problem necessarily,
unless the number is quite high (say, 500+ total threads).

Even though modern Ruby versions use kernel threads to back Ruby threads, and having more of those will increase the number of context switching and RAM used, for most Ruby-based system the different would be insignificant. To be honest, if you are that worried about consumer efficiency,
I'd suggest skipping Ruby entirely and using Go, JVM, Erlang or .NET.

Do you have any benchmarks  that demonstrate

 * The number of threads you use
 * How having more threads affects RAM usage
 * Ditto for CPU context switches for the Sneakers proces
 * How this affects, say, 95 percentile throughput and latency?

Consuming from multiple queues in a worker is not a difficult thing to do — just track them in a set and consume from/unsubscribe from each — but it does break the flow Sneakers users have gotten used to over the years. Sneakers by definition is an opinionated system with some limitations.

Consuming from N queues would also make reasoning about worker metrics harder because now you have to have a way to "average" metrics from N queues instead of knowing what single queue
to look at.

@jondot @gabrieljoelc WDYT?

**Comment by barbolo** [comment]
Just to be clear, this is exactly what I want to achieve in terms of architecture:

![image](https://user-images.githubusercontent.com/274526/27490950-aee4f02e-5816-11e7-9197-4a721e7a0626.png)

In my case, each job takes about 10 seconds to be consumed by a worker and uses limited resources (databases and external APIs). So I need to limit the total amount of active workers to match the available resources.

I'm planning to create a queue for each client of my service. Initially I'll have 30 queues, but may have hundreds/thousands in the future (the number of queues will match the number of clients).

Suppose that one of my clients sent 1 million jobs and after that all the other 29 clients send 100 jobs each. Thus, I have 1,002,900 jobs to process. If I do not load balance the workers across multiple queues, the other 29 clients will have to wait for the 1 million jobs of the first client to finish before receiving any results.

I believe the best way to solve this problem is to have a worker reading jobs from multiple queues.

**Comment by michaelklishin** [comment]
I'm not trying to say that this design is wrong (using a single queue for everything is) but if the issue is with downstream service capacity, a single runaway client will still cause a slowdown observed by other clients.

You probably have considered this but limiting queue length or using message TTL may be more effective solutions overall.

But thanks for the example, I'm more open to the idea now. Let's see what @jondot and @gabrieljoelc would say.

There are two other features in Sneakers and RabbitMQ can target the issue of variable and uneven load on consumers: [Autoscaling](https://github.com/jondot/sneakers/wiki/Auto-scaling) and [Federated Queues](http://www.rabbitmq.com/federated-queues.html).

**Comment by gabrieljoelc** [comment]
Thanks for checking out Sneakers @barbolo!

I agree with everything from @michaelklishin [comment](https://github.com/jondot/sneakers/issues/298#issuecomment-310704415). Using multiple queues per worker seems like more of a hack for the efficiency problem you are trying to solve. Sneakers is a framework for handling RabbitMQ messages in a _semantic_ way. The concurrency engine is what allows for working different workers in parallel. @michaelklishin suggestion to look at [Federated Queues](http://www.rabbitmq.com/federated-queues.html) could be a solution to solving it at the cluster level.

In conclusion, Sneakers should not support multiple queues per worker. You might need to consider @michaelklishin 's language advice to get the efficiency you are looking for:
> if you are that worried about consumer efficiency, I'd suggest skipping Ruby entirely and using Go, JVM, Erlang or .NET.

**Comment by seddy** [comment]
We were just looking for exactly this feature, but with a different use case, so I'd very much appreciate any thoughts you have on it @gabrieljoelc or @michaelklishin. 

We've got a number of services which are running Sneakers workers to consume messages from different queues on RabbitMQ using the [`MaxRetry` handler](https://github.com/jondot/sneakers/blob/master/lib/sneakers/handlers/maxretry.rb). 

Consequently, we have N apps, and 3N queues (`queue`, `queue-retry`, `queue-error`) and are struggling to stay on top of all the `*-error` queues; we have a dozen apps and counting which operate in this way...

What we were hoping to do was configure a single worker to process all the `*-error` queues in a single place. This would allow us to aggregate and monitor our errors in a separate app which will scale as our infrastructure grows without having to have one worker per `-error` queue. They are currently all on the same exchange but use different routing keys. 

Having a worker processing multiple queues would seem (naively perhaps!) to fit this use case.  Is there existing configuration we should be using instead? I took a look through the `Maxretry` example, code and specs but couldn't see anything appropriate; but this may well be my unfamiliarity with the code. There did seem to be reference to options such as [`retry-error-exchange`](https://github.com/jondot/sneakers/blob/master/lib/sneakers/handlers/maxretry.rb#L47) but I'm not sure how to specify the configuration in `Sneakers.configure`...

Any thoughts or suggestions very much appreciated 😄 !

**Comment by gabrieljoelc** [comment]
@seddy thanks for you interest in Sneakers!

It doesn't seem like you are looking for a semantic-driven worker framework. Are you really just looking for a helper/utility that allows you to subscribe one bit of code to multiple RabbitMQ queues? If yes and you are committed to using Ruby, I suggest you just familiarize yourself with the a Ruby RabbitMQ client (Sneakers uses [bunny](https://github.com/ruby-amqp/bunny)) and build a little utility to pull this off.

**Comment by michaelklishin** [comment]
This is not a commonly requested feature and it is indeed something that doesn't really fit into Sneakers' opinionated design. Closing, at least for now. Thank you for providing all the details to make your case, though. Thoughtful and detailed feature requests is a relatively rare thing to see on GitHub.
